{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "30970dc1",
   "metadata": {},
   "source": [
    "# Vectorized version of Gene Expression Programming\n",
    "- Here the genes are fully vectorized and tokenized\n",
    "- The algorithm just relies on three basis functions:\n",
    "- Dominant inheritance\n",
    "- Rezessiv inheritance\n",
    "- Gene Fussion\n",
    "- The programm mainly relies on numpy arrays for the genetic representation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "50baa348",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import some packages\n",
    "from deap.tools import selTournament\n",
    "from sklearn.metrics import mean_absolute_error, mean_squared_error\n",
    "from tqdm import tqdm\n",
    "import warnings\n",
    "import matplotlib.pyplot as plt\n",
    "from ipywidgets import interact, IntSlider\n",
    "import ipywidgets as widgets\n",
    "import numpy as np\n",
    "import numexpr\n",
    "import multiprocessing\n",
    "from multiprocessing import shared_memory\n",
    "import sympy as sp\n",
    "%matplotlib inline\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "np.random.seed(0)\n",
    "\n",
    "#Define some helper Functions\n",
    "def find_indices_with_sum(arr, target_sum, num_indices):\n",
    "    if arr[0]==-1:\n",
    "        return [0]\n",
    "    cum_sum = np.cumsum(arr)\n",
    "    indices = np.where(cum_sum == target_sum)[0]\n",
    "    if len(indices) >= num_indices:\n",
    "        return indices[:num_indices]\n",
    "    else:\n",
    "        return []\n",
    "    \n",
    "def compile_to_function_string(rek_string, arity_map, **kwargs):\n",
    "    \"\"\"\n",
    "    Using the compile recursive for the expression\n",
    "    \"\"\"\n",
    "    stack = []\n",
    "    for elem in rek_string[::-1]:\n",
    "        if arity_map.get(elem, 0) == 2:\n",
    "            op1 = stack.pop()\n",
    "            op2 = stack.pop()\n",
    "            stack.append('(' + op1 + elem + op2 + ')')\n",
    "        elif arity_map.get(elem, 0) == 1:\n",
    "            op1 = stack.pop()\n",
    "            stack.append(elem + '(' + op1 + ')')\n",
    "        elif arity_map.get(elem, 0) == -1:\n",
    "            op1 = stack.pop()\n",
    "            stack.append('(' + op1 + ')'+elem)\n",
    "        else:\n",
    "            stack.append(elem)\n",
    "    return stack[-1]\n",
    "\n",
    "\n",
    "#Class Toolbox - contains different info of the alg\n",
    "class Toolbox(object):\n",
    "    def __init__(self, gene_count, head_len, symbols, gene_connections, \n",
    "                 mutation_prob, crossover_prob, fusion_prob):\n",
    "        self.gene_count = gene_count\n",
    "        self.head_len = head_len\n",
    "        self.headsyms = {}\n",
    "        self.tailsyms = {}\n",
    "        self.id_to_char = {}\n",
    "        self.char_to_id = {}\n",
    "        self.arrity_by_id = {}\n",
    "        self.symbols = symbols\n",
    "        self.gene_connections = gene_connections\n",
    "        self.mutation_prob = mutation_prob\n",
    "        self.crossover = crossover_prob\n",
    "        self.fusion_prob = fusion_prob\n",
    "    \n",
    "    @property\n",
    "    def arrity_by_id(self):\n",
    "        return self._arrity_by_id\n",
    "    \n",
    "    @property\n",
    "    def fusion_prob(self):\n",
    "        return self._fusion_prob\n",
    "    \n",
    "    @fusion_prob.setter\n",
    "    def fusion_prob(self, fusion_prob):\n",
    "        self._fusion_prob = fusion_prob\n",
    "        return self\n",
    "    \n",
    "    @property\n",
    "    def crossover_prob(self):\n",
    "        return self._crossover_prob\n",
    "    \n",
    "    @property\n",
    "    def mutation_prob(self):\n",
    "        return self._mutation_prob\n",
    "    \n",
    "    @property\n",
    "    def gene_count(self):\n",
    "        return self._gene_count\n",
    "    \n",
    "    @property\n",
    "    def head_len(self):\n",
    "        return self._head_len\n",
    "    \n",
    "    @property\n",
    "    def headsyms(self):\n",
    "        return self._headsyms\n",
    "    \n",
    "    @property\n",
    "    def tailsyms(self):\n",
    "        return self._tailsyms\n",
    "    \n",
    "    @property\n",
    "    def symbols(self):\n",
    "        return self._symbols \n",
    "    \n",
    "    @property\n",
    "    def gene_connections(self):\n",
    "        return self._gene_connections\n",
    "    \n",
    "    @gene_connections.setter\n",
    "    def gene_connections(self, value): \n",
    "        self._gene_connections = [self.char_to_id[elem] for elem in value]\n",
    "    \n",
    "    @symbols.setter\n",
    "    def symbols(self, value):\n",
    "        self._symbols = value\n",
    "        self.char_to_id = {elem:index for index,elem in enumerate(value.keys())}\n",
    "        self.id_to_char = {index:elem for elem, index in self.char_to_id.items()}\n",
    "        self.headsyms = [self.char_to_id[key] for key,arity in  value.items()]\n",
    "        self.tailsyms = [self.char_to_id[key] for key,arity in  value.items() if arity<1]\n",
    "        self.arrity_by_id = {index:arity for (_,arity), index in zip(value.items(),\n",
    "                                                                     range(len(value.keys())))}\n",
    "        return self\n",
    "    \n",
    "    @head_len.setter\n",
    "    def head_len(self, value):\n",
    "        self._head_len = value\n",
    "        return self\n",
    "    \n",
    "    @gene_count.setter\n",
    "    def gene_count(self, value):\n",
    "        self._gene_count = value\n",
    "        return self\n",
    "    \n",
    "    @mutation_prob.setter\n",
    "    def mutation_prob(self, value):\n",
    "        self._mutation_prob = value\n",
    "        return self\n",
    "    \n",
    "    @crossover_prob.setter\n",
    "    def crossover_prob(self, value):\n",
    "        self._mutation_prob = value\n",
    "        return self\n",
    "    \n",
    "    @headsyms.setter\n",
    "    def headsyms(self, value):\n",
    "        self._headsyms = value\n",
    "        return self\n",
    "    \n",
    "    @tailsyms.setter\n",
    "    def tailsyms(self, value):\n",
    "        self._tailsyms = value\n",
    "        return self\n",
    "    \n",
    "    @arrity_by_id.setter\n",
    "    def arrity_by_id(self, value):\n",
    "        self._arrity_by_id = value\n",
    "        return self\n",
    "    \n",
    "#Repsents the genetic information\n",
    "class Chromosome(np.ndarray):\n",
    "    def __new__(cls, genes, toolbox=None):\n",
    "        obj = np.asarray(genes).view(cls)\n",
    "        obj.fitness = None\n",
    "        obj.toolbox = toolbox\n",
    "        obj.expression = None\n",
    "        obj.compile_expression()\n",
    "        return obj\n",
    "    \n",
    "    @property\n",
    "    def fitness(self):\n",
    "        return self._fitness\n",
    "    \n",
    "    @fitness.setter\n",
    "    def fitness(self, value,):\n",
    "        self._fitness = value\n",
    "        return self\n",
    "     \n",
    "    #Currently we deal at least with two genes\n",
    "    def _karva_raw(self):\n",
    "        gene_len = toolbox.head_len*2+1\n",
    "        connectionsym = self[:(toolbox.gene_count-1)]\n",
    "        genes = self[(toolbox.gene_count-1):]\n",
    "        arity_gene_ = np.vectorize(lambda x: toolbox.arrity_by_id[x])(genes)\n",
    "        arity_gene_ = np.where(arity_gene_==0, 0, arity_gene_)\n",
    "        rolled_indices = [connectionsym]\n",
    "        for i in range(0, len(arity_gene_) - gene_len + 1, gene_len):\n",
    "            window = arity_gene_[i:i + gene_len]\n",
    "            window[1:]=window[1:]-1\n",
    "            indices = find_indices_with_sum(window, 0, 1)\n",
    "            rolled_indices.append(genes[i:i+indices[0]+1])\n",
    "        return np.concatenate(rolled_indices)\n",
    "        \n",
    "    \n",
    "    def compile_expression(self):\n",
    "        if self.expression is None:\n",
    "            expression = self._karva_raw()\n",
    "            self.expression = [toolbox.id_to_char[elem] for elem in expression]\n",
    "            self.expression = compile_to_function_string(self.expression, self.toolbox.symbols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "03fef1b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Operator for generating\n",
    "def generate_gene(headsyms, tailsyms, headlen):\n",
    "    head=np.random.randint(np.min(headsyms),np.max(np.concatenate([headsyms, tailsyms], axis=0))+1,headlen)\n",
    "    tail=np.random.randint(np.min(tailsyms),np.max(tailsyms)+1,2*headlen+1)\n",
    "    return np.concatenate([head,tail])\n",
    "\n",
    "def generate_chromosome(toolbox):\n",
    "    headsyms = toolbox.headsyms\n",
    "    tailsyms = toolbox.tailsyms\n",
    "    head_len = toolbox.head_len\n",
    "    genecount = toolbox.gene_count\n",
    "    connectors = toolbox.gene_connections\n",
    "    gene_connectors = np.random.randint(\n",
    "        np.min(connectors),np.max(connectors),genecount-1)\n",
    "    genes = np.concatenate([generate_gene(headsyms, tailsyms, head_len) for _ in range(genecount)],axis=0) \n",
    "    return Chromosome(np.concatenate([gene_connectors,genes]), toolbox)\n",
    "      \n",
    "\n",
    "def generate_population(number, toolbox):\n",
    "    return [generate_chromosome(toolbox) for _ in range(number)]\n",
    "\n",
    "def create_operator_masks(gene_seq_alpha, gene_seq_beta, pb=0.2):\n",
    "    alpha_operator = np.zeros_like(gene_seq_alpha)\n",
    "    beta_operator = np.zeros_like(gene_seq_alpha)\n",
    "    indices_alpha = np.random.randint(0,len(gene_seq_alpha),min(int(pb*len(gene_seq_alpha)),len(gene_seq_alpha)))\n",
    "    indices_beta = np.random.randint(0,len(gene_seq_alpha),min(int(pb*len(gene_seq_alpha)),len(gene_seq_alpha)))\n",
    "    alpha_operator[indices_alpha] = 1\n",
    "    beta_operator[indices_beta] = 1\n",
    "    return alpha_operator, beta_operator\n",
    "\n",
    "    \n",
    "def gene_dominant_fusion(gene_seq_alpha, gene_seq_beta, pb=0.2):\n",
    "    alpha_operator, beta_operator = create_operator_masks(gene_seq_alpha, gene_seq_beta, pb)\n",
    "    child_1 = np.where(alpha_operator==1,np.maximum(gene_seq_alpha, gene_seq_beta),gene_seq_alpha) \n",
    "    child_2 = np.where(beta_operator==1,np.maximum(gene_seq_alpha, gene_seq_beta), gene_seq_beta)\n",
    "    return Chromosome(child_1, gene_seq_alpha.toolbox), Chromosome(child_2, gene_seq_alpha.toolbox)\n",
    "\n",
    "def gen_rezessiv(gene_seq_alpha, gene_seq_beta, pb=0.2):\n",
    "    alpha_operator, beta_operator = create_operator_masks(gene_seq_alpha, gene_seq_beta, pb)\n",
    "    child_1 = np.where(alpha_operator==1,np.minimum(gene_seq_alpha, gene_seq_beta),gene_seq_alpha) \n",
    "    child_2 = np.where(beta_operator==1,np.minimum(gene_seq_alpha, gene_seq_beta), gene_seq_beta)\n",
    "    return Chromosome(child_1, gene_seq_alpha.toolbox), Chromosome(child_2, gene_seq_alpha.toolbox)\n",
    "\n",
    "def gene_fussion(gene_seq_alpha, gene_seq_beta, pb=0.2):\n",
    "    alpha_operator, beta_operator = create_operator_masks(gene_seq_alpha, gene_seq_beta, pb)\n",
    "    child_1 = np.where(alpha_operator==1,(gene_seq_alpha+gene_seq_beta)//2,gene_seq_alpha) \n",
    "    child_2 = np.where(beta_operator==1,(gene_seq_alpha+gene_seq_beta)//2, gene_seq_beta)\n",
    "    return Chromosome(child_1, gene_seq_alpha.toolbox), Chromosome(child_2, gene_seq_alpha.toolbox)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1ffc4c94",
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate_fitness(prior_fit, expression, shared_mem_name, data_shape, feature_cols\n",
    "                     ,*,default_val=10e6, data_set_target='y'):\n",
    "    ret_val = default_val\n",
    "    \n",
    "    existing_shm = shared_memory.SharedMemory(name=shared_mem_name)\n",
    "    data_set = np.ndarray(data_shape, dtype=np.float64, buffer=existing_shm.buf)\n",
    "    data_set = {feature:data_set[:,col] for col,feature in enumerate(feature_cols)}\n",
    "    try:\n",
    "            pred = numexpr.evaluate(expression, local_dict=data_set)\n",
    "            ret_val = mean_absolute_error(data_set[data_set_target], pred)\n",
    "    except Exception as e:\n",
    "        ret_val = 10e6\n",
    "    return ret_val"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51632f0f",
   "metadata": {},
   "source": [
    "# Test the implementation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b3244c36",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a Dataset with 100 points\n",
    "x_data = np.linspace(-1,1,100)\n",
    "y_data = np.vectorize(lambda x: x**4+x**2+x+4)(x_data)\n",
    "data_set = np.stack([x_data,y_data], axis=1)\n",
    "features = ['x_0','y']\n",
    "data_shape = data_set.shape\n",
    "shm = shared_memory.SharedMemory(create=True, size=data_set.nbytes)\n",
    "data_set_shared = np.ndarray(data_shape, dtype=np.float64, buffer=shm.buf)\n",
    "data_set_shared[:] = data_set[:] \n",
    "\n",
    "#Define the symbols \n",
    "\n",
    "symbols = {\n",
    "    '+':2,\n",
    "    '*':2,\n",
    "    '-':2,\n",
    "    '/':2,\n",
    "    'x_0':0,\n",
    "    '2':0,\n",
    "}\n",
    "\n",
    "toolbox = Toolbox(3,10,symbols,{\"+\":2,'-':2},0.2,0.2,0.1)\n",
    "population = generate_population(1000,toolbox)\n",
    "error_function = mean_squared_error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cab7655e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  4%|█▊                                        | 21/500 [00:07<02:25,  3.30it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.13193266542924736\n",
      "Mean:423899.5622861359\n",
      "Min:0.13193266542924736\n",
      "Max:10000000.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  8%|███▍                                      | 41/500 [00:13<02:31,  3.02it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.13193266542924736\n",
      "Mean:110327.88840118525\n",
      "Min:0.13193266542924736\n",
      "Max:10000000.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 12%|█████                                     | 60/500 [00:20<02:36,  2.82it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.09182342461990022\n",
      "Mean:2502269799.6815095\n",
      "Min:0.09182342461990022\n",
      "Max:2501999792984.1646\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 15%|██████▎                                   | 75/500 [00:26<02:28,  2.85it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ready - function: x_0**4 + x_0**2 + x_0 + 4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "## Define the operators\n",
    "epochs = 500\n",
    "number_of_cores = 8\n",
    "mating_size = 0.75\n",
    "plot_data = []\n",
    "for epoch in tqdm(range(epochs)):\n",
    "    fits = []\n",
    "    with multiprocessing.Pool(processes=number_of_cores) as pool:\n",
    "        fitness_values = pool.starmap(evaluate_fitness, [(elem.fitness,elem.expression,shm.name, \n",
    "                                                          data_shape, features) for elem in population])\n",
    "\n",
    "    for elem, fitness in zip(population, fitness_values):\n",
    "        elem.fitness = fitness\n",
    "        fits.append(fitness)\n",
    "    \n",
    "    population = sorted(population, key=lambda x: x.fitness)\n",
    "    if epoch%20==0 and epoch>0:\n",
    "        print(population[0].fitness)\n",
    "        print(f\"Mean:{np.nanmean(fits)}\")\n",
    "        print(f\"Min:{np.nanmin(fits)}\")\n",
    "        print(f\"Max:{np.nanmax(fits)}\")\n",
    "        plot_data.append(population[0].expression)\n",
    "        \n",
    "    if epoch<epochs-2:\n",
    "        tournie = selTournament(population[:len(population)//2], int(len(population)*mating_size), 3, fit_attr='fitness')\n",
    "        next_gen = []\n",
    "        for index in range(0,len(tournie),2):\n",
    "                t1 = tournie[index]\n",
    "                t2 = tournie[(index+1)%len(tournie)]\n",
    "                if np.random.random()>0.5:\n",
    "                    t1,t2 = gene_dominant_fusion(t1,t2)\n",
    "                if np.random.random()>0.5:    \n",
    "                    t1,t2 = gen_rezessiv(t1,t2)\n",
    "                if np.random.random()>0.5:   \n",
    "                    t1,t2 = gene_fussion(t1,t2)\n",
    "                next_gen.append(t1)\n",
    "                next_gen.append(t2)\n",
    "        population = population[:int(len(population)*(1-mating_size))] + next_gen\n",
    "    if np.nanmin(fits)<1e-10:\n",
    "        print(f\"Ready - function: {sp.simplify(population[0].expression)}\")\n",
    "        plot_data.append(population[0].expression)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "84b2784e",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_set = {feature:data_set[:,col] for col,feature in enumerate(features)}\n",
    "plot_data_ = [(data_set['x_0'], data_set['y'], numexpr.evaluate(expression, local_dict=data_set), expression) \n",
    "              for expression in plot_data]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "503e5e76",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_expression(index):\n",
    "    x, y_true, y_pred, expr = plot_data_[index]\n",
    "    plt.figure(figsize=(8, 5))\n",
    "    plt.plot(x, y_true, label='True')\n",
    "    plt.plot(x, y_pred, label='Predicted')\n",
    "    plt.title(f\"Expression: {expr}\")\n",
    "    plt.legend()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3706b1dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a304c795ccd3468587492594a5177c0d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(IntSlider(value=0, description='index', max=3), Output()), _dom_classes=('widget-interac…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<function __main__.plot_expression(index)>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slider = IntSlider(min=0, max=len(plot_data_) - 1, step=1, value=0)\n",
    "interact(plot_expression, index=slider)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6db41922",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a3f4530",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
